### PR TITLE
Oracle price cache

### PR DESCRIPTION
**Pull request by Chepelau**

### CODE REVIEW

File: src/amm_core/oracles/agg.cairo [hunk 1]
```diff
@@ -1,25 +1,44 @@
...
+    use carmine_protocol::amm_core::state::State::{
+        read_latest_oracle_price, write_latest_oracle_price
+    };
+
     // @notice Returns current spot price for given ticker (quote and base token)
     // @param quote_token_addr: Address of quote token in given ticker
     // @param base_token_addr: Address of base token in given ticker
     // @return current_price: Current spot price
     fn get_current_price(
-        quote_token_addr: ContractAddress, base_token_addr: ContractAddress
```
Comment (Chepelau) [inline src/amm_core/oracles/agg.cairo:16]: why is this change necessary? this could break a lot of things no?

File: src/amm_core/oracles/agg.cairo [hunk 2]
```diff
@@ -1,25 +1,44 @@
...
-        price_pragma
+        let curr_block = get_block_info().unbox().block_number;
+        let (last_price, last_price_block_num) = read_latest_oracle_price(
+            base_token_addr, quote_token_addr
+        );
+
+        if last_price_block_num == curr_block {
+            last_price
+        } else {
+            let price_pragma = get_pragma_median_price(base_token_addr, quote_token_addr);
```
Comment (Chepelau) [inline src/amm_core/oracles/agg.cairo:31]:
```suggestion
            let price_pragma = get_pragma_median_price(quote_token_addr, base_token_addr);
```
This is the wrong way as well.

File: src/amm_core/oracles/agg.cairo [hunk 3]
```diff
@@ -1,22 +1,41 @@
...
+    //      the same block. This optimization reduces external calls and gas costs.
     // @param quote_token_addr: Address of quote token in given ticker
     // @param base_token_addr: Address of base token in given ticker
     // @return current_price: Current spot price
     fn get_current_price(
         quote_token_addr: ContractAddress, base_token_addr: ContractAddress
     ) -> Fixed {
-        let price_pragma = get_pragma_median_price(quote_token_addr, base_token_addr);
-        price_pragma
+        let curr_block = get_block_info().unbox().block_number;
```
Comment (tensojka) [inline src/amm_core/oracles/agg.cairo:26]: You can use get_block_number

**Review by tensojka — APPROVED** [review]
LGTM

File: src/amm_core/oracles/agg.cairo (hunk 3 above)
Comment (nickgore) [reply src/amm_core/oracles/agg.cairo:26]: apparently not in this version :shrug:

File: tests/test_fork_AMM.cairo [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+        .unwrap();
+    let amm = IAMMDispatcher { contract_address: amm_contract_addr };
+
+    let deployer: ContractAddress =
+        0x074fd7da23e21f0f0479adb435221b23f57ca4c32a0c68aad9409a41c27f3067
+        .try_into()
+        .unwrap();
+
+    let new_class_hash: ClassHash =
+        0x07a57625ddc2fbcca39c30a13c04fdcbc7d7c25ff7a66703666392dc0de1e88f
```
Comment (DaveVodrazka) [inline tests/test_fork_AMM.cairo:36]: class hash should be generated from the build, but we can fix that later on

**Comment by DaveVodrazka** [comment]
LGTM

**Review by tensojka — APPROVED** [review]
LGTM!
